### PR TITLE
zoekt-sourcegraph-indexserver: Instrument indexing prometheus

### PR DESCRIPTION
**Pull request by keegancsmith**
Part of https://github.com/sourcegraph/sourcegraph/issues/5471

### CODE REVIEW

**Review by kzh — APPROVED** [review]

**Comment by emidoots** [comment]
Thank you!!

**Comment by emidoots** [comment]
Here's some results from indexing 13k repos on v3.7.1 at a rate of 1417.75 repos/hr:

Average `resolveRevisionsDuration` (about 5-6s, not bad but also not great):

image

Average `resolveRevisionDuration` (under 1s, so most of the time is probably spent in `queue.AddOrUpdate`?):

image

And average time spent indexing repositories (this chart is interesting, most repositories are indexed in under a 3-20s, but some can take several minutes):

image

Ignore the errors at the end of the chart -- those come from 3 remaining repositories due to https://github.com/sourcegraph/sourcegraph/issues/5587

It would be good to also log the times of each repo we index, I bet there are some interesting outliers here we could improve.

**Comment by emidoots** [comment]
Actually, here is a chart showing the average indexing time with the average `resolveRevisionsDuration` time subtracted:

image

What this shows is that almost all of the "outlier" spikes are time spent indexing, while majority of the rest of the time is actually just spent resolving revisions.

**Comment by emidoots** [comment]
TL;DR:

- Optimizing how we resolve revisions would probably help speed up indexing a lot. 
- Logging the time spent indexing each repo would be a good idea, so we can catch outliers and investigate them.
- All of the above is nice-to-have, not critical.

**Comment by keegancsmith** [comment]
`resolveRevisionsDuration` is done once every 10minutes. So 3-5s is actually really good and indicates we can probably do listing more than once every 10min by default.

Also these are all histograms, so you should be using the histogram functions to extract useful numbers out of them. I don't think the measurements you are taking are the average, they are something else. To get the average you need something like

```prom
histogram_quantile(0.5, sum(rate(resolve_revisions_seconds_bucket[5m])) by (le))
```

Just so you can understand the flow of the indexer, there are two main goroutines:

### List goroutine
In a loop every 10min:
1. Fetch list of repositories to index
2. Find latest revision for each concurrently repository (`resolve_revisions_seconds` for all, `resolve_revision_seconds` for individual repositories)
3. Add/update the queue with the repo@commit.

Reason we track commit in the queue, is if the commit is different from what we have seen before, we put it in the front of the queue. Otherwise if we have already seen that commit (ie indexed it), we put it at the back of the queue.

### Index goroutine
In a loop:
1. If queue empty, sleep 10s
2. Pop front of queue
3. Index repo@commit (`index_repo_seconds`)
